plugins/osdn/vnids.go

```diff
@@ -9,22 +9,57 @@ import (
 
 	kapi "k8s.io/kubernetes/pkg/api"
 	kubetypes "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/registry/service/allocator"
+	etcdallocator "k8s.io/kubernetes/pkg/registry/service/allocator/etcd"
+	"k8s.io/kubernetes/pkg/util"
 	kerrors "k8s.io/kubernetes/pkg/util/errors"
 	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
-	"k8s.io/kubernetes/pkg/util/sets"
 	utilwait "k8s.io/kubernetes/pkg/util/wait"
 	"k8s.io/kubernetes/pkg/watch"
 
-	"github.com/openshift/openshift-sdn/pkg/netutils"
-	osapi "github.com/openshift/origin/pkg/sdn/api"
+	"github.com/openshift/openshift-sdn/pkg/netid"
+	vnidcontroller "github.com/openshift/openshift-sdn/plugins/osdn/netid/controller"
+	"github.com/openshift/openshift-sdn/plugins/osdn/netid/vnid"
+	"github.com/openshift/openshift-sdn/plugins/osdn/netid/vnidallocator"
+	"github.com/openshift/origin/pkg/sdn/registry/netnamespace"
 )
 
-const (
-	// Maximum VXLAN Network Identifier as per RFC#7348
-	MaxVNID = ((1 << 24) - 1)
-	// VNID for the admin namespaces
-	AdminVNID = uint(0)
-)
+func (oc *OsdnController) VnidStartMaster() error {
+	netIDRange, err := vnid.NewVNIDRange(netid.MinVNID, netid.MaxVNID-netid.MinVNID+1)
+	if err != nil {
+		return fmt.Errorf("Unable to create NetID range: %v", err)
+	}
+
+	var etcdAlloc *etcdallocator.Etcd
+	netIDAllocator := vnidallocator.New(netIDRange, func(max int, rangeSpec string) allocator.Interface {
+		mem := allocator.NewContiguousAllocationMap(max, rangeSpec)
+		etcdAlloc = etcdallocator.NewEtcd(mem, "/ranges/namespacevnids", kapi.Resource("namespacevnidallocation"), oc.EtcdHelper)
+		return etcdAlloc
+	})
+
+	_, kclient := oc.Registry.GetSDNClients()
+
+	// Run vnid migration
+	migrate := netnamespace.NewMigrate(oc.EtcdHelper, kclient.Namespaces())
+	if err := migrate.Run(); err != nil {
+		return fmt.Errorf("Unable to migrate network namespaces: %v", err)
```

Comment:
Yes, master will not start if migration fails here. If the problem persists (very rare) they could delete NetNamespace objects manually using oc delete or curl on etcd and SDN will generate and assign new network IDs. This will lose prior config (project network merges/enabled global network access). Should I add this message as part of the error?
Yes, We remove old NetNamespace objects as part of migration. In case of HA master, SDN controller runs on one of the master. So there is no issue during rolling update. Migration order for master or node doesn't matter but all the nodes need to be migrated as well. Old nodes will watch for NetNamespace and this resource no longer exists. This info needs to be documented in the release notes?